### PR TITLE
[Perf] Improve multimodal item handling from O(n) to O(log n) per step

### PR DESCRIPTION
**Pull request by andylolu2**
## Purpose

We observed that voxtral-realtime slows down a lot on long transcription sessions. The boils down to the fact that vLLM is not very efficient in handling a large amount of multimodal items. (For voxtral-realtime this can go up to 32K multimodal items).

### Proposed fix

Use bisect to narrow the iteration range over multimodal features in the scheduler and model runner from O(n) to O(log n). This significantly improves performance for models with many multimodal inputs per request.

Also maintain a per-request index in EncoderCacheManager to avoid scanning all mm_features in get_cached_input_ids(), and use torch.from_numpy instead of torch.tensor for zero-copy audio conversion (minor).

## Test Result

Profile before (voxtral-realtime with ~8k mm items):

Screenshot 2026-05-30 at 21 52 03

Profile after:

Screenshot 2026-05-30 at 21 51 06

Zooming into expensive parts.

Scheduler `.schedule()`:

600541179-696982b4-d323-482e-9ff7-78c9e1a3b3eb

GpuModelRunner `._preprocess()`:

600541203-aa75e95d-c021-40e6-a198-0c4263bdf3be

### CODE REVIEW

File: vllm/v1/core/encoder_cache_manager.py [hunk 1]
```diff
@@ -201,22 +205,13 @@ def allocate(self, request: Request, input_id: int) -> None:
...
-        """Get all cached multimodal input IDs for a request.
-
-        Returns the set of input IDs whose `mm_hash` exists in the cache map.
-        This includes entries that are currently unreferenced (and thus present
-        in `freeable`); for such entries, freeing for this request will be a
-        no-op.
-        """
-        return {
-            input_id
-            for input_id in range(len(request.mm_features))
```
Comment (andylolu2) [inline vllm/v1/core/encoder_cache_manager.py:217]: This would have been an O(n) loop every step otherwise

**Review by NickLucche — COMMENTED** [review]
Nice @andylolu2 !
Should we apply the same `get_mm_features_in_window` logic to Mrv2?

PS profile pics links are broken for me

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -1140,22 +1141,23 @@ def _try_schedule_encoder_inputs(
...
-        for i, mm_feature in enumerate(mm_features):
+
+        lo, hi = get_mm_features_in_window(
+            mm_features,
+            start=num_computed_tokens,
+            end=num_computed_tokens + num_new_tokens + shift_computed_tokens,
+        )
+        # For encoder-decoder, all inputs sit at start_pos=0, so lo=0 always.
+        if self.is_encoder_decoder:
+            lo = 0
```
Comment (andylolu2) [inline vllm/v1/core/sched/scheduler.py:1175]: @DarkLight1337 / @NickLucche I'm actually not sure if this special case is needed, but it is here before

**Review by ywang96 — APPROVED** [review]

File: vllm/v1/core/encoder_cache_manager.py [hunk 3]
```diff
@@ -234,6 +229,9 @@ def free_encoder_input(self, request: Request, input_id: int) -> None:
         if not self.cached.get(mm_hash, None):
             return
         self.cached[mm_hash].discard(req_id)
+        self.request_cached_ids[req_id].discard(input_id)
+        if not self.request_cached_ids[req_id]:
+            del self.request_cached_ids[req_id]
```
Comment (ywang96) [inline vllm/v1/core/encoder_cache_manager.py]: Now I think about it, should we move this cleanup up since there's an early return?

File: vllm/v1/core/encoder_cache_manager.py [hunk 4]
```diff
@@ -71,6 +71,8 @@ def __init__(self, cache_size: int):
 
         # mm_hash of mm_data => ids of requests that reference the mm_data
         self.cached: dict[str, set[str]] = {}
+        # request_id => set of input_ids cached for that request
+        self.request_cached_ids: dict[str, set[int]] = {}
```
Comment (njhill) [inline vllm/v1/core/encoder_cache_manager.py:75]: Use `defaultdict`?

File: vllm/v1/core/encoder_cache_manager.py (hunk 3 above)
Comment (andylolu2) [reply vllm/v1/core/encoder_cache_manager.py]: Yes we should actually, good catch!

Otherwise for a request with two identical mm items it would early return and `request_cached_ids` won't be cleanly returned. I've made the fix and added a test.

**Review by njhill — APPROVED** [review]
Thanks @andylolu2 , nice improvement!

File: vllm/v1/core/encoder_cache_manager.py (hunk 4 above)
Comment (andylolu2) [reply vllm/v1/core/encoder_cache_manager.py:75]: I think using dict is a tiny bit safer because if we use a defaultdict just doing `return request_cached_ids[req_id]` would create a new empty entry in `request_cached_ids`. I think not needing to worry about these edge cases is bit more comforting.

Comment (andylolu2) [reply vllm/v1/core/encoder_cache_manager.py:75]: (and the "cost" is just two `setdefault` calls)
